File: app/src/lib/attachments.ts
import { invoke, convertFileSrc } from '@tauri-apps/api/core';
import { join } from '@tauri-apps/api/path';

const ALLOWED_EXT = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'svg', 'heic'];

let cachedAppDataDir: string | null = null;

async function getAppDataDir(): Promise<string> {
  if (cachedAppDataDir) return cachedAppDataDir;
  cachedAppDataDir = await invoke<string>('get_app_data_dir');
  return cachedAppDataDir;
}

function pickExt(file: File | Blob, fallbackName?: string): string {
  const name = (file as File).name || fallbackName || '';
  const fromName = name.includes('.') ? name.split('.').pop()!.toLowerCase() : '';
  if (ALLOWED_EXT.includes(fromName)) return fromName;
  const fromMime = (file.type || '').split('/').pop()?.toLowerCase() ?? '';
  if (ALLOWED_EXT.includes(fromMime)) return fromMime;
  return 'png';
}

/** 上传图片：File/Blob → 写入 app_data_dir/attachments → 返回相对路径 "attachments/xxx.ext" */
export async function uploadImage(file: File | Blob, fallbackName?: string): Promise<string> {
  const ext = pickExt(file, fallbackName);
  const buf = await file.arrayBuffer();
  const bytes = Array.from(new Uint8Array(buf));
  return await invoke<string>('save_attachment', { ext, bytes });
}

/** 笔记里的相对路径 → webview 能加载的 asset URL */
export async function resolveAttachmentUrl(relPath: string): Promise<string> {
  // 已经是 url（http/https/blob/data）就直接返回
  if (/^(https?:|data:|blob:|asset:)/.test(relPath)) return relPath;
  const dir = await getAppDataDir();
  const abs = await join(dir, relPath);
  return convertFileSrc(abs);
}

export function isImageFile(file: File | Blob): boolean {
  return file.type.startsWith('image/');
}

// 匹配 markdown 里的相对路径引用：attachments/xxx.ext（含中文/字母/数字/-/_/.）
const ATTACHMENT_REF_RE = /attachments\/[A-Za-z0-9._-]+/g;

/** 从一组笔记 markdown 里提所有被引用的相对路径 */
export function extractAttachmentRefs(contents: string[]): string[] {
  const set = new Set<string>();
  for (const text of contents) {
    const matches = text.match(ATTACHMENT_REF_RE);
    if (matches) matches.forEach(m => set.add(m));
  }
  return Array.from(set);
}

/** 清理孤儿附件：传入「所有被引用的相对路径」，后端删除其余文件（60s 内修改的跳过） */
export async function cleanupOrphans(referenced: string[]): Promise<number> {
  return await invoke<number>('cleanup_orphan_attachments', { referenced });
}

File: src-tauri/src/attachments.rs
use std::collections::HashSet;
use std::path::Path;
use std::time::SystemTime;

use tauri::{Manager, State};

use crate::db::Db;

/// 保存附件到 {app_data_dir}/attachments/{uuid}.{ext}
/// 返回相对路径 "attachments/{uuid}.{ext}"，前端写入 markdown。
#[tauri::command]
pub fn save_attachment(
    app: tauri::AppHandle,
    ext: String,
    bytes: Vec<u8>,
) -> Result<String, String> {
    let safe_ext = ext.trim_start_matches('.').to_lowercase();
    let allowed = ["png", "jpg", "jpeg", "gif", "webp", "bmp", "svg", "heic"];
    if !allowed.contains(&safe_ext.as_str()) {
        return Err(format!("不支持的图片格式: {}", safe_ext));
    }

    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e| e.to_string())?
        .join("attachments");
    std::fs::create_dir_all(&dir).map_err(|e| e.to_string())?;

    let filename = format!("{}.{}", uuid::Uuid::new_v4(), safe_ext);
    let full_path = dir.join(&filename);
    std::fs::write(&full_path, &bytes).map_err(|e| e.to_string())?;

    Ok(format!("attachments/{}", filename))
}

/// 给前端用：返回 app_data_dir 的绝对路径，渲染时拼接相对路径再转成 webview URL。
#[tauri::command]
pub fn get_app_data_dir(app: tauri::AppHandle) -> Result<String, String> {
    app.path()
        .app_data_dir()
        .map(|p| p.to_string_lossy().into_owned())
        .map_err(|e| e.to_string())
}

/// 删除 attachments/ 下没被任何笔记引用的孤儿文件。
#[tauri::command]
pub fn cleanup_orphan_attachments(
    app: tauri::AppHandle,
    db: State<'_, Db>,
) -> Result<usize, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e| e.to_string())?
        .join("attachments");

    if !dir.exists() {
        return Ok(0);
    }

    let referenced_files = referenced_attachment_files(&db)?;
    let entries = std::fs::read_dir(&dir).map_err(|e| e.to_string())?;
    let mut deleted = 0usize;

    for entry in entries.flatten() {
        let path = entry.path();
        if !path.is_file() {
            continue;
        }
        let Some(name) = path.file_name().and_then(|n| n.to_str()) else {
            continue;
        };
        if referenced_files.contains(name) {
            continue;
        }
        if modified_less_than_60_seconds_ago(&entry) {
            continue;
        }
        if trash::delete(&path).is_ok() {
            deleted += 1;
        }
    }

    Ok(deleted)
}

fn referenced_attachment_files(db: &Db) -> Result<HashSet<String>, String> {
    let conn = db.conn.lock().map_err(|e| e.to_string())?;
    let mut stmt = conn
        .prepare("SELECT content_md FROM notes")
        .map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map([], |row| row.get::<_, String>(0))
        .map_err(|e| e.to_string())?;

    let mut refs = HashSet::new();
    for text in rows.filter_map(Result::ok) {
        collect_attachment_refs(&text, &mut refs);
    }

    Ok(refs
        .into_iter()
        .filter_map(|p| {
            Path::new(&p)
                .file_name()
                .and_then(|n| n.to_str())
                .map(|s| s.to_string())
        })
        .collect())
}

fn collect_attachment_refs(text: &str, refs: &mut HashSet<String>) {
    const PREFIX: &str = "attachments/";
    let mut rest = text;

    while let Some(pos) = rest.find(PREFIX) {
        let matched = &rest[pos..];
        let mut end = 0usize;
        for (idx, ch) in matched.char_indices() {
            if idx < PREFIX.len() || ch.is_ascii_alphanumeric() || matches!(ch, '.' | '_' | '-') {
                end = idx + ch.len_utf8();
            } else {
                break;
            }
        }
        if end > PREFIX.len() {
            refs.insert(matched[..end].to_string());
        }
        rest = &matched[end.max(PREFIX.len())..];
    }
}

fn modified_less_than_60_seconds_ago(entry: &std::fs::DirEntry) -> bool {
    entry
        .metadata()
        .and_then(|meta| meta.modified())
        .ok()
        .and_then(|modified| SystemTime::now().duration_since(modified).ok())
        .is_some_and(|elapsed| elapsed.as_secs() < 60)
}
